### PR TITLE
write from POSTGRES to Google Sheets

### LINKED ISSUE #96: write from postgres to google sheets
Top answer here is good:
https://stackoverflow.com/questions/9690138/how-do-i-access-read-write-google-sheets-spreadsheets-with-python
But even better is this video.
https://www.youtube.com/watch?v=4ssigWmExak&ab_channel=LearnGoogleSpreadsheets

### PR DESCRIPTION
**Pull request by oliverangelil**

### CODE REVIEW

**Comment by oliverangelil** [comment]
@RicSegundo I pretty much copied this excellent video: https://www.youtube.com/watch?v=4ssigWmExak&ab_channel=LearnGoogleSpreadsheets. Am writing to [this](https://docs.google.com/spreadsheets/d/12kzUd8wHKWDomBz0M2ng-6zQ_t46UblKiSnMebD5su4/edit#gid=0) google sheets. 

There is also [gspread](https://github.com/burnash/gspread) which is a wrapper around the sheets API, but I prefer to directly use the API like in the video. 

Things that need to be improved:
- there is a `creds.json` file in the repo that the API reads to connect to the sheet. I tried to add it as a JSON to github secrets but that threw an error in the deploy workflow (couldn't be passed as an environment variable). Also I recall reading somewhere that JSON should not be passed as a secret. 
- there's no column names in google sheets after writing. 
- I am inheriting a class here, however there's the wise saying "composition over inheritance": https://python-patterns.guide/gang-of-four/composition-over-inheritance/

**Comment by oliverangelil** [comment]
In the creds.json file there is an email address that one must give editor access to in the google sheets
image

**Comment by RicSegundo** [comment]
I'll check asap, in principle this weekend!

**Comment by oliverangelil** [comment]
https://stackoverflow.com/questions/56277661/is-it-possible-to-store-a-json-file-to-an-env-variable-with-dotenv
https://stackoverflow.com/questions/71544103/how-can-we-store-a-json-credential-to-env-variable-in-python
https://medium.com/personal-work-revision/how-to-decode-base64-encoded-88b2c08ba64c

**Comment by oliverangelil** [comment]
@RicSegundo I solved 2 of the 3 "areas for improvement" listed above:
- pass the json through github secrets as base64, and then decode back to json on the VM
- write the column names to google sheets

Hope you can support on the last bullet (the general design of the code). Also need to add type hinting...

**Comment by oliverangelil** [comment]
I have removed `sonar-project.properties` file and replaced it with `.sonarcloud.properties`. `sonar-project.properties` is only used for the CI approach but we're using the automatic approach (recommended)

File: src/scraping/results.py [hunk 1]
```diff
@@ -29,13 +28,13 @@ def extract_results(
             print("Cannot save to database without pre-processing the results.")
             return
         # save results to database
-        db = T.DataBaseInteraction(results, D.DatabaseTables.TABLE_ghana_2022.value)
-        db.save_results_to_db(db_type)
+        db = T.DataBaseInteraction()
```
Comment (RicSegundo) [inline src/scraping/results.py]: I think that properties of the object should be passed at the time of creation of the object, and not when you call the method (and then it saves to self.something)

File: src/scraping/tools.py [hunk 2]
```diff
@@ -1,12 +1,16 @@
+import os
+import base64
 import re
 import requests
 import numpy as np
 import pandas as pd
 import json
 import functools
 import sqlalchemy
+from googleapiclient.discovery import build
```
Comment (RicSegundo) [inline src/scraping/tools.py]: import the whole namespace, and call build from it below

File: src/scraping/tools.py [hunk 3]
```diff
@@ -182,23 +186,21 @@ class DataBaseInteraction:
     table_name (str): name of the table to be created/used in the database
     """
 
-    def __init__(
-        self,
-        dataframe: pd.DataFrame,
-        table_name: str
-                ) -> None:
-
+    def __init__(self) -> None:
```
Comment (RicSegundo) [inline src/scraping/tools.py]: re-add data frame and table_name, and set dataframe to  = None so it is not mandatory to pass when instantiating the object

File: src/scraping/tools.py [hunk 4]
```diff
@@ -182,23 +186,21 @@ class DataBaseInteraction:
...
         self.db_path: str = D.DatabaseConnection.DB_PATH
         self.host: str = D.DatabaseConnection.HOST
         self.user: str = C.Postgres.USER
         self.password: str = C.Postgres.PASS
         self.port: str = D.DatabaseConnection.PORT
         self.db_name: str = D.DatabaseConnection.DB_NAME
-        self.dataframe: pd.DataFrame = dataframe
-        self.table_name: str = table_name
-        self.db_engine: sqlalchemy.engine.base.Engine = None
+        self.db_engine: sqlalchemy.engine.base.Engine = None,
```
Comment (RicSegundo) [inline src/scraping/tools.py]: change the order of these 2 lines because I am OCD

File: src/scraping/tools.py [hunk 5]
```diff
@@ -182,23 +186,21 @@ class DataBaseInteraction:
...
+        self.db_type: str = D.DatabaseTypes.POSTGRES
 
-    def save_results_to_db(self, db_type: str = D.DatabaseTypes.SQLITE) -> None:
+    def save_results_to_db(
+        self,
+        dataframe: pd.DataFrame,
```
Comment (RicSegundo) [inline src/scraping/tools.py]: remove this as unnecessary now

File: src/scraping/tools.py [hunk 6]
```diff
@@ -210,7 +212,8 @@ def save_results_to_db(self, db_type: str = D.DatabaseTypes.SQLITE) -> None:
         """
 
         # Create a connection engine
-        self.db_type = db_type
+        self.dataframe = dataframe
```
Comment (RicSegundo) [inline src/scraping/tools.py]: remove as unnecessary

File: src/scraping/tools.py [hunk 7]
```diff
@@ -253,3 +256,52 @@ def dataframe_to_db(self) -> None:
...
                     )
+
+
+class GoogleSheets(DataBaseInteraction):
+    def __init__(self) -> None:
+        super().__init__()
+
+    @staticmethod
+    def base64_to_json(b64) -> dict[str, str]:
+        decodedbytes: bytes = base64.b64decode(b64[1:-1])
```
Comment (RicSegundo) [inline src/scraping/tools.py]: I'll send the link to the named ranges and list slicing, e.g. https://www.w3schools.com/python/ref_func_slice.asp

File: src/scraping/tools.py [hunk 8]
```diff
@@ -253,3 +256,52 @@ def dataframe_to_db(self) -> None:
...
+        decodedstr: str = decodedbytes.decode("ascii")
+        json_dict: dict[str, str] = json.loads(decodedstr)
+        return json_dict
+
+    def read_from_sql(self, table_name: str) -> None:
+        self.db_connect()
+        df: pd.DataFrame = pd.read_sql(table_name, con=self.db_engine)
```
Comment (RicSegundo) [inline src/scraping/tools.py]: maybe create another method that process the data?

Comment (RicSegundo) [reply src/scraping/tools.py]: @oliverangelil had the great idea of checking the column type and making it automatic to change the date formats

File: src/scraping/tools.py [hunk 9]
```diff
@@ -253,3 +256,52 @@ def dataframe_to_db(self) -> None:
...
+        # create service account credentials object
+        creds = service_account.Credentials.from_service_account_info(json_dict, scopes=SCOPES)
+
+        # Construct a Resource for interacting with an API
+        service = build('sheets', 'v4', credentials=creds)
+
+        # convert dataframe into list, with first list being column names
+        data: list[list[Any]] = self.dataframe.to_numpy().tolist()
+        column_names: list[str] = self.dataframe.columns.tolist()
+        data.insert(0, column_names)
```
Comment (RicSegundo) [inline src/scraping/tools.py]: maybe give a name to the 0, so we know what it is? put it in a variable?

Comment (RicSegundo) [reply src/scraping/tools.py]: maybe you can put data = column_names (already as a list of lists with only 1 list) + data

File: src/scraping/tools.py [hunk 10]
```diff
@@ -253,3 +256,52 @@ def dataframe_to_db(self) -> None:
...
+
+        # instantiate class to interact with a resource
+        sheet = service.spreadsheets()
+
+        # write to google sheets
+        result: dict[str, Any] = sheet.values().update(
+            spreadsheetId="12kzUd8wHKWDomBz0M2ng-6zQ_t46UblKiSnMebD5su4",
+            range="test!A1",
```
Comment (RicSegundo) [inline src/scraping/tools.py]: maybe divide this into 3 actions and not cascade them?

File: src/scraping/tools.py [hunk 11]
```diff
@@ -253,3 +256,52 @@ def dataframe_to_db(self) -> None:
...
+        df.replace(np.nan, 'N/A', inplace=True)
+        self.dataframe: pd.DataFrame = df
+
+    def writetosheets(self) -> dict[str, Any]:
+        SCOPES: list[str] = ['https://www.googleapis.com/auth/spreadsheets']
+        b64: str = os.environ['SHEETS_API_CREDENTIALS_B64']
+        json_dict: dict[str, str] = self.base64_to_json(b64)
+
+        # create service account credentials object
+        creds = service_account.Credentials.from_service_account_info(json_dict, scopes=SCOPES)
```
Comment (RicSegundo) [inline src/scraping/tools.py]: add the correct name spacing here

File: src/scraping/tools.py (hunk 9 above)
Comment (RicSegundo) [reply src/scraping/tools.py]: (maybe add this to the data pre-processing)?

File: src/scraping/tools.py (hunk 10 above)
Comment (RicSegundo) [reply src/scraping/tools.py]: and I know you know, write all the constants into the definition files
(every time I see literal string text here a dolphin dies)

**Review by RicSegundo — CHANGES_REQUESTED** [review]

**Comment by oliverangelil** [comment]
@RicSegundo, the code works perfectly, but there is a mypy issue that I cannot resolve. 
It complains about this line:
`super()._DataBaseInteraction__db_connect()`

This SO question explains the syntax:
https://stackoverflow.com/a/20261595/5392289

we can change the method from private to protected (`_db_connect()`)
and then call `super()._db_connect()`
That would work... maybe mypy is complaining for good reason??

**Comment by oliverangelil** [comment]
@RicSegundo it should be protected. I've fixed it. You can approve.
https://stackoverflow.com/a/12784107/5392289

**Review by RicSegundo — APPROVED** [review]

**Pull request merged**